devimint/src/external.rs

```diff
@@ -240,11 +242,90 @@ impl Bitcoind {
         Ok(proof.encode_hex())
     }
 
-    pub async fn get_raw_transaction(&self, txid: bitcoin30::Txid) -> Result<String> {
+    /// Poll a transaction by its txid until it is found in the mempool or in a
+    /// block.
+    pub async fn poll_get_transaction(&self, txid: bitcoin30::Txid) -> anyhow::Result<String> {
+        poll("Waiting for transaction in mempool", || async {
+            match self
+                .get_transaction(txid)
+                .await
+                .context("getrawtransaction")
+            {
+                Ok(Some(tx)) => Ok(tx),
+                Ok(None) => Err(ControlFlow::Continue(anyhow::anyhow!(
+                    "Transaction not found yet"
+                ))),
+                Err(err) => Err(ControlFlow::Break(err)),
+            }
+        })
+        .await
+    }
+
+    /// Get a transaction by its txid. Checks the mempool and all blocks.
+    async fn get_transaction(&self, txid: bitcoin30::Txid) -> Result<Option<String>> {
+        // Check the mempool.
+        match self.get_raw_transaction(txid, None).await {
+            // The RPC succeeded, and the transaction was not found in the mempool. Continue to
+            // check blocks.
+            Ok(None) => {}
+            // The RPC failed, or the transaction was found in the mempool. Return the result.
+            other => return other,
+        };
+
+        let block_height = self.get_block_count().await? - 1;
+
+        // Check each block for the tx, starting at the chain tip.
+        // Buffer the requests to avoid spamming bitcoind.
+        // We're doing this after checking the mempool since the tx should
+        // usually be in the mempool, and we don't want to needlessly hit
+        // the bitcoind with block requests.
+        let mut buffered_tx_stream = futures::stream::iter((0..block_height).rev())
+            .map(|height| async move {
+                let block_hash = self.get_block_hash(height).await?;
+                self.get_raw_transaction(txid, Some(block_hash)).await
+            })
+            .buffered(32);
+
+        while let Some(tx_or) = buffered_tx_stream.next().await {
+            match tx_or {
+                // The RPC succeeded, and the transaction was not found in the block. Continue to
+                // the next block.
+                Ok(None) => continue,
+                // The RPC failed, or the transaction was found in the block. Return the result.
+                other => return other,
+            };
+        }
+
+        Err(anyhow!("Transaction not found"))
```

Comment:
I think this needs to return `Ok(None)` for the retry logic to work.

Reply:
Whoops! Good call, looks like that fixed CI

Reply:
Just verified that this fixes the devimint tests